### ISSUE TITLE
query-frontend: Response caching should work for subqueries and part queries.

### ISSUE THREAD

**Issue by bwplotka** [cortexproject/cortex#2178]
AC:
* Subqueries should be split & cached.

I think that should not be that hard to implement, but we would need to finally parse the query and understand what is nested.

cc @brancz, @pracucci @tomwilkie

**Comment by pracucci** [comment]
@owen-d This is your area of expertise. What's your take, especially considering the recently merged #1878 ?

**Comment by bwplotka** [comment]
We are happy to help in free time, but ideas/suggestions are very welcome.

Or if any work happened already for this (:

**Comment by bwplotka** [comment]
Any update on this? :hugs: 

Any pointers for contributors, e.g if we would want to tackle it?

**Comment by owen-d** [comment]
I'm not sure if this will be straightforward to implement. We currently sequence a set of middlewares in the query frontend, of which a caching middleware is one. This makes it hard to pull out part of a query without dealing with the promql Engine, which is what we did in the sharding work (https://github.com/cortexproject/cortex/pull/1878). This could be another reason to support a refactoring of the query frontend though. There are a number of issues derived from sequencing middlewares this way, such as https://github.com/cortexproject/cortex/issues/1882

Caching middleware: https://github.com/cortexproject/cortex/blob/master/pkg/querier/queryrange/results_cache.go

Another thought - this is something we may want to solve via some sort of query planner.

**Comment by bwplotka** [comment]
Yes, definitely worth to look closer in to different design then. I would
say initial promql parse might be necessary...

**Comment by bwplotka** [comment]
Something to add as well : https://github.com/thanos-io/thanos/issues/2569

**Comment by pracucci** [comment]
_Still valid_

**Comment by pracucci** [comment]
_Still valid_

**Comment by pracucci** [comment]
_Still valid_

**Comment by jtlisi** [comment]
_still valid_

**Comment by bboreham** [comment]
Definitely still valid. 

Example: you can write an instant query like `count(up)[60m:1m]` which is basically the same thing as a range-query of `count(up)` over 60m with step 1m.  Query-frontend will cache the latter but not the former.
